### PR TITLE
feat(bedrock): agent action groups

### LINKED ISSUE #169: (new construct): Implement Amazon Bedrock features as CDK constructs
### Describe the feature

Amazon Bedrock recently unveiled some features which are not available through CDK at the moment (Guardrails, knowledge bases, agents). 

### Use Case

Provide new features for end users

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by jstrunk**
Fixes #169 

Depends on #231 and #232 

Adds an AgentActionGroup custom resource and construct.

Usage example:
```ts
  const agent = new bedrock.Agent(this, 'Agent', {
    foundationModel: bedrock.BedrockFoundationModel.ANTHROPIC_CLAUDE_INSTANT_V1_2,
    instruction: 'You are a helpful and friendly agent that answers questions about literature.',
    knowledgeBases: [kb],
  });

  const actionGroupFunction = new lambda_python.PythonFunction(this, 'ActionGroupFunction', {
    runtime: lambda.Runtime.PYTHON_3_12,
    entry: path.join(__dirname, '../lambda/action-group'),
    layers: [lambda.LayerVersion.fromLayerVersionArn(this, 'PowerToolsLayer', `arn:aws:lambda:${this.region}:017000801446:layer:AWSLambdaPowertoolsPythonV2:60`)],
  });


agent.addActionGroup({
  actionGroupName: 'query-library',
  description: 'Use these functions to get information about the books in the library.',
  actionGroupExecutor: actionGroupFunction,
  actionGroupState: "ENABLED",
  apiSchema: bedrock.ApiSchema.fromAsset(path.join(__dirname, 'action-group.yaml')),
});
```

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

**Comment by massi-ang** [comment]
Looks good to me. I would like to have  an `ApiSchema` construct , similar to the `Code` construct, with methods `from_file()` and `from_inline()`, instead of using `fs.ReadFileSync()` directly. 

Expose also a method `addActionGroup` on the `Agent` construct. 
Example on user experience:

```
const agent = new Agent(this, "weather-agent", {
      instruction:
        "You are a weather expert and answer user question about weather in different places. You answer the questions in the same language they have been asked.",
      description: "an agent to interact with a weather api",
    });

    agent.addActionGroup({
      lambda: weather,
      apiSchema: {
        payload: ApiSchema.fromPath(
          path.join(__dirname, "weather", "schema.json")
        ),
      },
      actionGroupState: "ENABLED",
    });

```

**Comment by jstrunk** [comment]
> Here are some suggestion for UX improvements:
> 
>     1. Add  an `ApiSchema` construct , similar to the `Code` construct, with methods `fromPath()` and `fromInline()`, instead of using `fs.ReadFileSync()` directly. This would also hide the language specific implementations for reading files.
> 
>     2. Expose a method `addActionGroup()` on the `Agent` construct.
 
Thank you @massi-ang ! I like these ideas! I'll implement them.

**Comment by jstrunk** [comment]
> Here are some suggestion for UX improvements:
> ```
> 1. Add  an `ApiSchema` construct , similar to the `Code` construct, with methods `fromPath()` and `fromInline()`, instead of using `fs.ReadFileSync()` directly. This would also hide the language specific implementations for reading files.
> 
> 2. Expose a method `addActionGroup()` on the `Agent` construct.
> ```

@massi-ang `fromPath()` is `fromAsset()` to be consistent with `Code`.

File: test/cdk-lib/bedrock/agent-action-group.test.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        description: 'Use these functions to get information about the books in the Project Gutenburg library.',
+        parentActionGroupSignature: 'AMAZON.UserInput',
+        actionGroupExecutor: actionGroupFunction,
+        actionGroupState: 'ENABLED',
+        apiSchema: bedrock.ApiSchema.fromInline(actionGroupAPISpec),
+      });
+    }).toThrowError('Cannot specify both description and parentActionSignature');
+  });
+
+  // test('Fails when both apiSchema.payload and apiSchema.s3 are provided', () => {
```
Comment (hvital) [inline test/cdk-lib/bedrock/agent-action-group.test.ts:104]: Is this commented code block still needed?

Comment (jstrunk) [reply test/cdk-lib/bedrock/agent-action-group.test.ts:104]: Thanks for catching that. After adding the ApiSchema class, I don't have a way to test that. I'm removing it.

**Comment by jstrunk** [comment]
The sample app in https://github.com/jstrunk/generative-ai-cdk-constructs-samples/tree/bedrock-agent-action-groups has been updated with a function that queries the most popular books in the library.

**Review by hvital — APPROVED** [review]

**Review by krokoko — APPROVED** [review]

**Pull request merged**